src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -144,8 +179,101 @@ CoordinatorInsertSelectExecScanInternal(CustomScanState *node)
+			Job *distSelectJob = distSelectPlan->workerJob;
+			List *distSelectTaskList = distSelectJob->taskList;
+			TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+			bool randomAccess = true;
+			bool interTransactions = false;
+			bool binaryFormat =
+				CanUseBinaryCopyFormatForTargetList(selectQuery->targetList);
+
+			ExecuteSubPlans(distSelectPlan);
+
+			/*
+			 * We have a separate directory for each transaction, so choosing
+			 * the same result prefix won't cause filename conflicts. Results
+			 * directory name also includes node id and database id, so we don't
+			 * need to include them in the filename. Jobs are executed
+			 * sequentially, so we also don't need to include job id here.
+			 */
+			char *distResultPrefix = "repartitioned_results";
+
+			DistTableCacheEntry *targetRelation =
+				DistributedTableCacheEntry(targetRelationId);
+
+			int partitionColumnIndex =
+				PartitionColumnIndex(insertTargetList, targetRelation->partitionColumn);
+			if (partitionColumnIndex == -1)
+			{
+				char *relationName = get_rel_name(targetRelationId);
+				Oid schemaOid = get_rel_namespace(targetRelationId);
+				char *schemaName = get_namespace_name(schemaOid);
+
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg(
+									"the partition column of table %s should have a value",
+									quote_qualified_identifier(schemaName,
+															   relationName))));
+			}
+
+			TargetEntry *selectPartitionTE = list_nth(selectQuery->targetList,
+													  partitionColumnIndex);
+			const char *partitionColumnName = selectPartitionTE->resname ?
+											  selectPartitionTE->resname : "(none)";
+
+			ereport(DEBUG2, (errmsg(
+								 "partitioning SELECT query by column index %d with name %s",
+								 partitionColumnIndex, quote_literal_cstr(
+									 partitionColumnName))));
+
+			int targetColumnCount = list_length(insertTargetList);
+			List **redistributedResults = RedistributeTaskListResults(distResultPrefix,
+																	  distSelectTaskList,
+																	  targetColumnCount,
+																	  partitionColumnIndex,
+																	  targetRelation,
+																	  binaryFormat);
+
+			/*
+			 * At this point select query has been executed on workers and results
+			 * have been fetched in such a way that they are colocated with corresponding
+			 * target shard. Create and execute a list of tasks of form
+			 * INSERT INTO ... SELECT * FROM read_intermediate_results(...);
+			 */
+			List *taskList = RedistributedInsertSelectTaskList(insertSelectQuery,
+															   targetRelation,
+															   redistributedResults,
+															   binaryFormat);
+
+			scanState->tuplestorestate =
+				tuplestore_begin_heap(randomAccess, interTransactions, work_mem);
+
+			TransactionProperties xactProperties = {
+				.errorOnAnyFailure = true,
+				.useRemoteTransactionBlocks = TRANSACTION_BLOCKS_REQUIRED,
+				.requires2PC = false
```

Comment:
Should this be true?

Reply:
Changed to using `ExecuteTaskListIntoTupleStore()` which decides on the correct value without explicitly assigning it.